### DISCUSSION TITLE
MutableDict marking changes when value is unchanged

### DISCUSSION THREAD

**Discussion by AbdealiLoKo** [sqlalchemy/sqlalchemy#7970] [Usage Questions] [answered]
### Describe the bug

When I use MutableDict in a column, it tracks changes while I modify the dict.
But even if my modification is back to the original value - it still detects that as a change

### To Reproduce

```python
>>> from sqlalchemy import inspect
>>> from models import MyModel
>>> item = MyModel.query.first()

>>> item.info
{'key': 1}
>>> inspect(item).attrs.info.history
History(added=(), unchanged=[{'key': 1}], deleted=())

>>> item.info['key'] = 2
>>> inspect(item).attrs.info.history
History(added=[{'key': 2}], unchanged=(), deleted=())

>>> item.info['key'] = 1
>>> inspect(item).attrs.info.history
History(added=[{'key': 1}], unchanged=(), deleted=())
```

I would have expected that the last `inspect` would give unchanged.
This works fine for Integer/String/other types

```python

>>> inspect(item).attrs.version.history
History(added=(), unchanged=[1], deleted=())

>>> item.version = 2
>>> inspect(fv).attrs.version.history
History(added=[1.1], unchanged=(), deleted=[1])

>>> item.version = 1
>>> inspect(fv).attrs.version.history
History(added=(), unchanged=[1], deleted=())
```

### Error

None

### Versions

- OS: Windows 10 + Ubuntu 18.04 WSL2
- Python: 3.7.4
- SQLAlchemy: 1.4.35
- Database: sqlite3
- DBAPI: sqlite3

### Additional context

I am actually not using `MutableDict` but have my own class. I am trying to understand how this should be handled for my own Mutable class also.

**Comment by zzzeek** [comment, marked as answer]
MutableDict is mutated in place, so there is no "previous" dictionary for it to compare itself towards.  therefore it's not possible for a change back to the "original" value to be detected since this case modifies that same "original" value.

**Comment by AbdealiLoKo** [comment]
@zzzeek interesting. That seems to make sense, but creates some odd usage.

Can I create my own variant of Dict which can handle this better ? Is there a recipe that I could leverage ?

Maybe I could keep track of the 'initial value' and if the 'current value' is different from 'initial value' the history comes cleanly ?

**Reply by zzzeek** [reply]
you can make a dictionary subclass that intercepts change events and then records history in its own internal way, and then consume that later, but that would not be through the SQLAlchemy API.

or you could use an immutable form of dictionary and simply assign a new dictionary to the attribute rather than mutating it.   this would be the way to do this cleanly.  SQLAlchemy makes great use of immutable dictionaries internally.

**Comment by AbdealiLoKo** [comment]
I wouldnt mind recording the history myself in my own dictionary subclass.
But is there a way for me to send that back to sqlalchemy so that I can use it with inspect() ?

I have no issue in actually tracking the added/unchanged/deleted in my own class and then just send that info to swlalchrmy every time a change occurs.

**Reply by zzzeek** [reply]
inspect gives you the loaded values of the attribute:

```py
inspect(item).attrs.version.value.my_history
```

**Comment by AbdealiLoKo** [comment]
But that would be a custom attribute that anyone using the inspect API would need to be aware of.

For example - if some other library I use is using `inspect(item).attrs.version.has_changes()` to identify which columns in the DB has been modified - I guess that would not work ?

**Reply by zzzeek** [reply]
it seems like you are building a 3rd party tool of some kind and you'd like it to be smoothly integrated for users.    

so, if you want to do that, you need to give the history API what it needs, which is that when someone updates the dictionary, there is an "old" value of that dictionary (can be any object really) that when compared to the "new" dictionary, it says, "yes, this has changed" or "no, this has not changed".

you want to do this without actually replacing the dictionary on the object.  otherwise, this would be easy as it's the default behavior.


so look inside of the `Mutable` base class here:  https://github.com/sqlalchemy/sqlalchemy/blob/b51a5f596299e1e52ca27b2db20394110f584257/lib/sqlalchemy/ext/mutable.py#L530-L535

flag_modified() is the function here setting up that something changed:

https://github.com/sqlalchemy/sqlalchemy/blob/b51a5f596299e1e52ca27b2db20394110f584257/lib/sqlalchemy/orm/attributes.py#L2242-L2263

what we want to do is to be able to put this "old" object, that will explicitly compare as "yes it's changed" or "no it's not changed" into that API call, like this:

```py
def flag_modified(instance, key, old_value=NO_VALUE):
    state, dict_ = instance_state(instance), instance_dict(instance)
    impl = state.manager[key].impl
    impl.dispatch.modified(state, impl._modified_token)
    state._modified_event(dict_, impl, old_value, is_userland=True)
```

that's a proposed API that could happen if we first confirm this works.   iterim approach is to do this:

```py
class MyMutable(Mutable, ...):
    # ...

   def changed(self):

        for parent, key in self._parents.items():
            object = parent.obj()
            flag_modified(object, key)
            inspect(object).commited_state[key] = MySpecialCompareThing(use_this_dictionary_to_compare=self)
```

note I'm proposing a change-detecting object that has a reference to the thing you're tracking.  maybe just the `__eq__()` method on your dictionary can do a special thing when it receives `MySpecialCompareThing` as an argument.

anyway, that value inside of the `committed_state` dictionary controls if history shows you "added" or "unchanged".   that's where you need to put a key.  if you can work out a clean example and get it all working, we can add an argument to flag_modified() to allow this use case.
